Snowplow migration doc (#955)
rudderlabs/rudderstack-docs: docs/user-guides/migration-guides/snowplow-migration-guide.mdx

---
title: "Migrating from Snowplow to RudderStack"
description: Step-by-step guide on migrating from Snowplow to RudderStack.
---

> This document covers the necessary steps to migrate from [Snowplow](https://snowplow.io/) to RudderStack and replace your instrumentation code. Once done, you can start using the RudderStack SDKs to track your events with minimal code changes.

Reviewer: ```suggestion
This document covers the necessary steps to migrate from [Snowplow](https://snowplow.io/) to RudderStack and replace your instrumentation code. You can then start using the RudderStack SDKs to track your events with minimal code changes.
```

## Setting up RudderStack dashboard

> 1. [Sign up](https://app.rudderstack.com/signup) for the RudderStack dashboard.

Reviewer: ```suggestion
1. [Sign up](https://app.rudderstack.com/signup) for RudderStack Cloud.
```

2. Set up the source-destination connections in the dashboard to facilitate the event data flow. For more information on setting up the connections, refer to the <Link to="/get-started/quickstart/">Quickstart</Link> guide.

<div class="successBlock">
Note the <Link to="/resources/glossary/#write-key">source write key</Link> and the <Link to="/dashboard-guides/overview/#data-plane-url">data plane URL</Link> as they will be used in the next section.
</div>

## Updating the SDK implementation

Depending on the platform, follow these steps to move your existing Snowplow SDK implementation to RudderStack:

<Tabs>
  <TabList>
    <Tab>Android</Tab>
    <Tab>iOS</Tab>
    <Tab>JavaScript</Tab>
  </TabList>
    <TabPanels>
      <TabPanel>
        Follow the below steps to migrate to the RudderStack Android SDK:

<div class="infoBlock">
        RudderStack distributes the Rudder Snowplow adapter SDK through <a href="https://search.maven.org/">Maven Central</a>. It is recommended to add the SDK to your project through the Android Gradle build system.
        </div>
        
      <ol>

>       <li>Add the following lines of code in your project level <code class="inline-code">app/build.gradle</code> file:

Reviewer: ```suggestion
      <li>Add the following lines of code in your project-level <code class="inline-code">app/build.gradle</code> file:
```

Reviewer: @pallabmaiti if we're saying "project-level", does including only `build.gradle` as opposed to `app/build.gradle` make more sense?

<span>

```groovy
buildscript {
  repositories {
    mavenCentral()
  }
}
allprojects {
  repositories {
    mavenCentral()
  }
}
```
</span>
</li>
<li>Add the following dependency under <code class="inline-code">dependencies</code> section in your <code class="inline-code">app/build.gradle</code> file:
<span>

```groovy
implementation 'com.rudderstack.android.snowplow:adapter:1.0.0.beta.1'
```
</span>
</li>

> <li>Update your SDK initialization to any one of the following snippet (method 1, 2, or 3). Also, replace the <code class="inline-code">WRITE_KEY</code> and <code class="inline-code">DATA_PLANE_URL</code> with your source write key and data plane URL obtained in the above section.

Reviewer: Took me some time to figure out where these methods were, hence the addition.

```javascript
<script>
  rs=window.rs=[],rs.snoplowAdapter=function(){rs.push(Array.prototype.slice.call(arguments))},rs.snoplowAdapter("newTracker",<WRITE_KEY>,<DATA_PLANE_URL>,{<Configurations>});
  //Optionally, use RudderStack JavaScript SDK load options
  rudderanalytics.load();
</script>

<script src="https://cdn.rudderlabs.com/adapters/v1/rs-sp-adapter.min.js"></script>
```
The <code class="inline-code">&lt;Configurations&gt;</code> parameter in the above snippet can be replaced by the following Snowplow properties which are mapped to the RudderStack properties as shown:

| Snowplow property| RudderStack property    | 
| :---------- | :------------ | 
| cookieDomain      |        setCookieDomain      |           
| cookieSameSite      |   sameSiteCookie            |           
| cookieSecure     |   secureCookie            |              

</TabPanel>
  </TabPanels>
</Tabs>

## Setting the configuration objects for mobile SDKs

> RudderStack supports setting values for the following Snowplow configuration objects in the mobile SDKs (iOS/Android). If not set, the default values are assigned.

Reviewer: What are the default values in this case? Should we clarify this bit? @pallabmaiti

Reviewer 2: Then we need to explain a lot.

Reviewer 2: For NetworkConfiguration,
dataPlaneUrl: https://hosted.rudderlabs.com
controlPlaneUrl: https://api.rudderlabs.com
SessionConfiguration:
backgroundTimeout: 5 mins
TrackerConfiguration:
Already present in the doc.

Reviewer 2: I don't know how you can fit this information there.

<h3>NetworkConfiguration</h3>

> The <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/configuring-how-events-are-sent/"><code class="inline-code">NetworkConfiguration</code></a> class can be used to specify the collector endpoint as shown:

Reviewer: ```suggestion
The <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/configuring-how-events-are-sent/"><code class="inline-code">NetworkConfiguration</code></a> class can be used to specify the collector endpoint, as shown:
```

    <Tab>Java</Tab>
    <Tab>Kotlin</Tab>
  </TabList>
    <TabPanels>
        <TabPanel>
<span>

```objectivec
// 1
RSNetworkConfiguration *networkConfig = [[RSNetworkConfiguration alloc] initWithDataPlaneUrl:DATA_PLANE_URL];
// 2
RSNetworkConfiguration *networkConfig = [[RSNetworkConfiguration alloc] initWithDataPlaneUrl:DATA_PLANE_URL controlPlaneUrl:CONTROL_PLANE_URL];
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```swift
// 1
let networkConfig = NetworkConfiguration(dataPlaneUrl: DATA_PLANE_URL)
// 2
let networkConfig = NetworkConfiguration(dataPlaneUrl: DATA_PLANE_URL, controlPlaneUrl: CONTROL_PLANE_URL)
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```java
// 1
NetworkConfiguration networkConfiguration = new NetworkConfiguration(DATA_PLANE_URL);
// 2
NetworkConfiguration networkConfiguration = new NetworkConfiguration(DATA_PLANE_URL, CONTROL_PLANE_URL);
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```kotlin
// 1
val networkConfiguration = NetworkConfiguration(DATA_PLANE_URL)
// 2
val networkConfiguration = NetworkConfiguration(DATA_PLANE_URL, CONTROL_PLANE_URL)
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>
<h3>SessionConfiguration</h3>

> The <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/#session"><code class="inline-code">SessionConfiguration</code></a> class can be used to capture sessions which help to track the user activity in the app as shown:

Reviewer: ```suggestion
The <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/#session"><code class="inline-code">SessionConfiguration</code></a> class can be used to capture sessions to track the user activity in the app, as shown:
```

```
</span>
      </TabPanel>
      <TabPanel>
<span>

```swift
// 1
let sessionConfig = SessionConfiguration(
  foregroundTimeout: Measurement(value: 30, unit: .minutes),
  backgroundTimeout: Measurement(value: 30, unit: .minutes)
)

// 2
let sessionConfig = SessionConfiguration(foregroundTimeoutInSeconds: 60, backgroundTimeoutInSeconds: 60)
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```java
SessionConfiguration sessionConfiguration = new SessionConfiguration(
  new TimeMeasure(1, TimeUnit.MINUTES),
  new TimeMeasure(1, TimeUnit.MINUTES)
);
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```kotlin
val sessionConfiguration = SessionConfiguration(
  TimeMeasure(1, TimeUnit.MINUTES),
  TimeMeasure(1, TimeUnit.MINUTES)
);
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>

<div class="warningBlock">
You need to pass both the arguments for <code class="inline-code">SessionConfiguration</code> class as shown in the above code snippets. However, RudderStack ignores the first argument (as it is a dummy argument) and considers only the second argument (<code class="inline-code">backgroundTimeout</code>).
</div>

<div class="infoBlock">

> For mobile SDKs, Snowplow's default timeout session is 30 minutes whereas RuddeStack's default timeout session is 5 minutes. Refer to the <Link to="/sources/event-streams/sdks/session-tracking/#mobile-sdks">RudderStack Session Tracking</Link> documentation for more information.

Reviewer: ```suggestion
For mobile SDKs, Snowplow's default timeout session is 30 minutes whereas RudderStack's default timeout session is 5 minutes. Refer to the <Link to="/sources/event-streams/sdks/session-tracking/#mobile-sdks">RudderStack Session Tracking</Link> documentation for more information.
```

</div>   

<h3>TrackerConfiguration</h3>

> The <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/"><code class="inline-code">TrackerConfiguration</code></a> class can be used to configure contexts and automatic events of the tracker, and general behaviour as shown:

Reviewer: ```suggestion
The <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/"><code class="inline-code">TrackerConfiguration</code></a> class can be used to configure contexts and automatic events of the tracker and the general behavior, as shown in the below snippets:
```

  .appId(APP_ID);
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```kotlin
val trackerConfiguration = TrackerConfiguration()
  .base64Encoding(false)
  .logLevel(LogLevel.VERBOSE)
  .deepLinkContext(true)
  .applicationContext(true)
  .platformContext(true)
  .geoLocationContext(true)
  .lifecycleAutotracking(true)
  .diagnosticAutotracking(true)
  .screenViewAutotracking(true)
  .screenContext(true)
  .applicationContext(true)
  .exceptionAutotracking(true)
  .installAutotracking(true)
  .userAnonymisation(false)
  .appId(APP_ID)
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>

Snowplow automatically captures and tracks the following data. Refer to the <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/#auto-tracked-events-and-entities">Auto-tracked events and entities</a> section for more information.

| Variable name | Default value     | 
| :---------- | :------------ | 
| logLevel     |        LogLevel.OFF       |           
| lifecycleAutotracking      |   true            |           
| screenViewAutotracking     |   true            |           
| sessionContext      |         true      |   

> 

Reviewer: Should we style the above variable names and their default values in backticks, i.e., code in text?

<div class="warningBlock">
RudderStack ignores any variable other than the ones mentioned above.
</div>

<h3>SubjectConfiguration</h3>

> The <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/"><code class="inline-code">SubjectConfiguration</code></a> class can be used to capture basic information about the user which is attached on all events as context entity.

Reviewer: ```suggestion
The <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/"><code class="inline-code">SubjectConfiguration</code></a> class can be used to capture basic user information which is attached to all the events as the context entity.
```

<div class="warningBlock">

> Snowplow does not provide any call to identify a user. Although, RudderStack sends an <Link to="/event-spec/standard-events/identify/"><code class="inline-code">identify</code></Link> call if you initialize the <code class="inline-code">SubjectConfiguration</code> class while initializing the SDK. If not initialized, a user is not identified. 

Reviewer: ```suggestion
Snowplow does not provide any call to identify a user. However, RudderStack sends an <Link to="/event-spec/standard-events/identify/"><code class="inline-code">identify</code></Link> call if you initialize the <code class="inline-code">SubjectConfiguration</code> class while initializing the SDK. <strong>Note that a user is not identified if the SDK is not initialized</strong>. 
```

</div>

The <code class="inline-code">userId</code> field is mapped to the RudderStack's <code class="inline-code">userId</code>. All the other properties are mapped to RudderStack's <code class="inline-code">traits</code> object.

<div class="warningBlock">

> Note that <code class="inline-code">userId</code> is a mandatory field. If not provided, the <code class="inline-code">identify</code> call is ignored.

Reviewer: ```suggestion
<code class="inline-code">userId</code> is a mandatory field. If not provided, the <code class="inline-code">identify</code> call is ignored.
```

| `LogLevel` | `LogLevel` | `LogLevel` | `LogLevel` | `LogLevel` |

## Sending event data

You can migrate your existing events from Snowplow to RudderStack as described below:

### Identifying a user

<Tabs>
  <TabList>
    <Tab>iOS/Android</Tab>
    <Tab>JavaScript</Tab>
  </TabList>
    <TabPanels>
      <TabPanel>
Snowplow's <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/"><code class="inline-code">SubjectConfiguration</code></a> class can be used to identify the users. Refer to the <Link to="/user-guides/migration-guides/snowplow-migration-guide/#setting-the-configuration-objects">Setting the configuration objects</Link> section for more information.
      </TabPanel>
      <TabPanel>

Snowplow supports identifying a user by passing the <code class="inline-code">setUserId</code> method. A Snowplow event including the <code class="inline-code">setUserId</code> method triggers the <Link to="/event-spec/standard-events/identify"><code class="inline-code">identify</code></Link> call in the RudderStack JavaScript SDK.

<div class="infoBlock">

> RudderStack supports only the <code class="inline-code">setUserId</code> method of the Snowplow v2 JS tracker.

Reviewer: Should we add some reference / link to the v2 JS tracker?

</div>

> In the below sample call, <code class="inline-code">alex@example.com</code> is the <code class="inline-code">userId</code> and the <code class="inline-code">firstName</code>, <code class="inline-code">lastName</code>, and <code class="inline-code">city</code> are the user traits.

Reviewer: ```suggestion
In the below sample call, <code class="inline-code">alex@example.com</code> is the <code class="inline-code">userId</code>, while <code class="inline-code">firstName</code>, <code class="inline-code">lastName</code>, and <code class="inline-code">city</code> are the user traits.
```

<span>

```javascript
rs.snowplowAdapter('setUserId', 'alex@example.com', {
  firstName: 'Alex',
  lastName: 'Keener',
  city: 'New Orleans',
});
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>

### Tracking user actions

RudderStack SDKs support the following types of Snowplow events:

<Tabs>
  <TabList>
    <Tab>iOS/Android</Tab>
    <Tab>JavaScript</Tab>
  </TabList>
    <TabPanels>
      <TabPanel>
<h3>Custom structured events</h3>

> RudderStack sends a <Link to="/event-spec/standard-events/track/">track</Link> call for Snowplow events containing the <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/#creating-a-structured-event"><code class="inline-code">Structured</code></a> class.

Reviewer: ```suggestion
RudderStack sends a <Link to="/event-spec/standard-events/track/">track</Link> call for the Snowplow events containing the <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/#creating-a-structured-event"><code class="inline-code">Structured</code></a> class.
```

> In the below example, RudderStack maps <code class="inline-code">Action_example</code> to RudderStack event name and the rest of the properties, like <code class="inline-code">Category_example</code>, <code class="inline-code">label</code>, <code class="inline-code">value</code>, etc. to the RudderStack properties.

Reviewer: ```suggestion
In the below example, RudderStack maps <code class="inline-code">Action_example</code> to the RudderStack event name and the rest of the properties like <code class="inline-code">Category_example</code>, <code class="inline-code">label</code>, <code class="inline-code">value</code>, etc. to the RudderStack properties.
```

  .label("my-label")
  .property("my-property")
  .value(5)
  .properties(["key_1": "value_1", "key_2": "value_2"])

tracker.track(structured)
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```java
HashMap < String, Object > properties = new HashMap < > ();
properties.put("key-1", "value 1");
properties.put("key-2", 123);
properties.put("key-3", 123.45);
properties.put("key-4", true);
properties.put("key-5", false);

Structured structured = new Structured("Category_example", "Action_example")
  .label("my-label")
  .value(1234.23)
  .property("my-property")
  .properties(properties);

tracker.track(structured);
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```kotlin
val properties = mapOf < String, Any > (
  "key-1"
  to "value 1",
  "key-2"
  to 123,
  "key-3"
  to 123.45,
  "key-4"
  to true,
  "key-5"
  to false
)

val structured = Structured("Category_example", "Action_example")
  .label("my-label")
  .value(1234.23)
  .property("my-property")
  .properties(properties)

tracker.track(structured)
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>

<h3>Custom self described events</h3>

> RudderStack sends a <Link to="/event-spec/standard-events/track/">track</Link> call for Snowplow events containing the <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/#creating-a-structured-event"><code class="inline-code">SelfDescribing</code></a> class.

Reviewer: ```suggestion
RudderStack sends a <Link to="/event-spec/standard-events/track/">track</Link> call for the Snowplow events containing the <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/#creating-a-structured-event"><code class="inline-code">SelfDescribing</code></a> class.
```

// 2
let selfDescribing = SelfDescribing(schema: "schema", payload: ["action": "Action_2"])

tracker.track(selfDescribing)
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```java
HashMap < String, Object > properties = new HashMap < > ();
properties.put("action", "Action-2");
properties.put("key-1", "value 1");

// 1
SelfDescribingJson selfDescribingJson = new SelfDescribingJson("schema", payload)
SelfDescribing selfDescribing = new SelfDescribing(selfDescribingJson)

// 2
SelfDescribing selfDescribing = new SelfDescribing("schema", payload)

tracker.track(selfDescribing);
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```kotlin
val payload = mapOf < String, Any > (
  "action"
  to "Action-2",
  "key-2"
  to "value-2"
)

// 1
val selfDescribingJson = SelfDescribingJson("schema", payload)
val selfDescribing = SelfDescribing(selfDescribingJson)

// 2
val selfDescribing = SelfDescribing("schema", payload)

tracker.track(selfDescribing)
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>

<div class="infoBlock">
<code class="inline-code">action</code> is a mandatory field. RudderStack does not send any call if it is absent.
</div>

<h3>Custom foreground events</h3>

> RudderStack sends a <Link to="/event-spec/standard-events/track/">track</Link> call for Snowplow events containing the <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/previous-versions/objective-c-tracker/objective-c-1-0-0/#foreground-and-background-events"><code class="inline-code">Foreground</code></a> class.

Reviewer: ```suggestion
RudderStack sends a <Link to="/event-spec/standard-events/track/">track</Link> call for the Snowplow events containing the <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/previous-versions/objective-c-tracker/objective-c-1-0-0/#foreground-and-background-events"><code class="inline-code">Foreground</code></a> class.
```

> RudderStack sends the event name as <code class="inline-code">Application Opened</code> and maps the rest of the properties, like <code class="inline-code">index</code>, <code class="inline-code">properties</code>, etc. to the RudderStack properties.

Reviewer: ```suggestion
RudderStack sends the event name as <code class="inline-code">Application Opened</code> and maps the rest of the properties like <code class="inline-code">index</code>, <code class="inline-code">properties</code>, etc. to the RudderStack properties.
```

  </TabList>
    <TabPanels>
        <TabPanel>
<span>

```objectivec
RSForeground * foreground = [
  [RSForeground alloc] initWithIndex: @1
];
[foreground properties: @ {
  @ "key_1": @ "value_1"
}];

[tracker track: foreground];
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```swift
let foreground = Foreground(index: 1)
  .properties(["key_1": "value_1"])

tracker.track(foreground)
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```java
HashMap < String, Object > properties = new HashMap < > ();
properties.put("key-1", "value 1");
properties.put("key-2", 123);
properties.put("key-3", 123.45);
properties.put("key-4", true);
properties.put("key-5", false);

Foreground foreground = new Foreground(1234);
foreground.setProperties(properties);

tracker.track(foreground);
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```kotlin
val properties = mapOf < String, Any > (
  "key-1"
  to "value 1",
  "key-2"
  to 123,
  "key-3"
  to 123.45,
  "key-4"
  to true,
  "key-5"
  to false
)

Foreground foreground = Foreground(1234)
foreground.properties = properties

tracker.track(background)
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>

<div class="infoBlock">

> This is not an automatic lifecycle event. Hence, the properties like <code class="inline-code">version</code>, <code class="inline-code">build</code>, etc. are not present under <code class="inline-code">properties</code>.

Reviewer: "automatic lifecycle event" could be confusing IMO. Should we instead say "default application lifecycle event" and link it to our lifecycle events doc? @pallabmaiti @ishachopra0401

</div>

<h3>Custom background events</h3>

> RudderStack sends a <Link to="/event-spec/standard-events/track/">track</Link> call for Snowplow events containing the <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/previous-versions/objective-c-tracker/objective-c-1-0-0/#foreground-and-background-events"><code class="inline-code">Background</code></a> class.

Reviewer: ```suggestion
RudderStack sends a <Link to="/event-spec/standard-events/track/">track</Link> call for the Snowplow events containing the <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/previous-versions/objective-c-tracker/objective-c-1-0-0/#foreground-and-background-events"><code class="inline-code">Background</code></a> class.
```

  </TabList>
    <TabPanels>
        <TabPanel>
<span>

```objectivec
RSBackground * background = [
  [RSBackground alloc] initWithIndex: @1
];
[background properties: @ {
  @ "key_1": @ "value_1"
}];

[tracker track: background];
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```swift
let background = Background(index: 1)
  .properties(["key_1": "value_1"])

tracker.track(background)
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```java
HashMap < String, Object > properties = new HashMap < > ();
properties.put("key-1", "value 1");
properties.put("key-2", 123);
properties.put("key-3", 123.45);
properties.put("key-4", true);
properties.put("key-5", false);

Background background = new Background(1234);
background.setProperties(properties);

tracker.track(background);
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```kotlin
val properties = mapOf < String, Any > (
  "key-1"
  to "value 1",
  "key-2"
  to 123,
  "key-3"
  to 123.45,
  "key-4"
  to true,
  "key-5"
  to false
)

Background background = Background(1234)
background.properties = properties

tracker.track(background)
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>

<div class="infoBlock">

> This is not an automatic lifecycle event. Hence, the properties like <code class="inline-code">version</code>, <code class="inline-code">build</code>, etc. are not present under <code class="inline-code">properties</code>.

Reviewer: Same comment as above.

Reviewer 2: Yes.

</div>
  </TabPanel>
      <TabPanel>

>         RudderStack JavaScript SDK supports the Snowplow's <code class="inline-code">trackStructEvent</code> and <code class="inline-code">trackSelfDescribingEvent</code> calls. They capture user events along with their associated properties and trigger the <Link to="/event-spec/standard-events/track/"><code class="inline-code">track</code></Link> call in RudderStack JavaScript SDK.

Reviewer: ```suggestion
        The RudderStack JavaScript SDK supports the Snowplow's <code class="inline-code">trackStructEvent</code> and <code class="inline-code">trackSelfDescribingEvent</code> calls. These Snowplow calls capture user events along with their associated properties and trigger the <Link to="/event-spec/standard-events/track/"><code class="inline-code">track</code></Link> call in the RudderStack JavaScript SDK.
```

Reviewer: Current sentence has some scope for confusion. Let me know if this tweak is accurate.

  <TabList>
    <Tab>Java</Tab>
    <Tab>Kotlin</Tab>
  </TabList>
    <TabPanels>
      <TabPanel>
<span>

```java
// 1
NetworkConfiguration networkConfiguration = new NetworkConfiguration(DATA_PLANE_URL);
// 2
NetworkConfiguration networkConfiguration = new NetworkConfiguration(DATA_PLANE_URL, CONTROL_PLANE_URL);
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```kotlin
// 1
val networkConfiguration = NetworkConfiguration(DATA_PLANE_URL)
// 2
val networkConfiguration = NetworkConfiguration(DATA_PLANE_URL, CONTROL_PLANE_URL)
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>
<h3>SessionConfiguration</h3>

The <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/#session"><code class="inline-code">SessionConfiguration</code></a> class can be used to capture the sessions which help to track of the user activity in the app as shown:

<Tabs>
  <TabList>
    <Tab>Java</Tab>
    <Tab>Kotlin</Tab>
  </TabList>
    <TabPanels>
      <TabPanel>
<span>

```java
SessionConfiguration sessionConfiguration = new SessionConfiguration(
  new TimeMeasure(1, TimeUnit.MINUTES),
  new TimeMeasure(1, TimeUnit.MINUTES)

> );

Reviewer 3: Actually, we are internally ignoring the first argument i.e., first argument is a dummy argument

> new SessionConfiguration(
>   ~new TimeMeasure(1, TimeUnit.MINUTES),~
>   new TimeMeasure(1, TimeUnit.MINUTES)
> );

Although the user needs to pass both the arguments but internally we will only consider the second argument.

- So, can you please add a info block at the appropriate place which mention this info.
- Can we add a reference to our (Mobile) Session Tracking doc. It'll could help in understanding the session behaviour properly. (Actually, Snowplow default timeout session is of 30 minutes in case of mobile but we have only of 5 minutes for mobile. Hence I want the first time user to navigate to the proper doc).

Author: @1abhishekpandey for the first point, is it applicable only for the Android SDK as I see different arguments for the iOS SDK?

Reviewer 3: It is applicable for both Android and iOS.

In iOS we've this code: 
> RSSessionConfiguration *sessionConfig = [[RSSessionConfiguration alloc] initWithForegroundTimeout: ~[[NSMeasurement alloc] initWithDoubleValue:30 unit:NSUnitDuration.minutes]~ backgroundTimeout:[[NSMeasurement alloc] initWithDoubleValue:30 unit:NSUnitDuration.minutes]];

In short, for mobile we'll only consider `backgroundTimeout` and it is the second argument in both cases (Android & iOS).

```
</span>
      </TabPanel>
      <TabPanel>
<span>

```kotlin
val sessionConfiguration = SessionConfiguration(
  TimeMeasure(1, TimeUnit.MINUTES),
  TimeMeasure(1, TimeUnit.MINUTES)
);
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>

<h3>TrackerConfiguration</h3>

> The <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/"><code class="inline-code">TrackerConfiguration</code></a> class can be used to capture information about the device and the app as shown:

Reviewer 2: This is similar to `RudderConfigBuilder`
Can you change the description?

  .userAnonymisation(false)
  .appId(APP_ID);
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```kotlin
val trackerConfiguration = TrackerConfiguration()
  .base64Encoding(false)
  .logLevel(LogLevel.VERBOSE)
  .deepLinkContext(true)
  .applicationContext(true)
  .platformContext(true)
  .geoLocationContext(true)
  .lifecycleAutotracking(true)
  .diagnosticAutotracking(true)
  .screenViewAutotracking(true)
  .screenContext(true)
  .applicationContext(true)
  .exceptionAutotracking(true)
  .installAutotracking(true)
  .userAnonymisation(false)
  .appId(APP_ID)
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>

Snowplow automatically captures and tracks the following data. Refer to the <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/#auto-tracked-events-and-entities">Auto-tracked events and entities</a> section for more information.

| Variable name | Default value     | 
| :---------- | :------------ | 
| logLevel     |        LogLevel.OFF       |           
| lifecycleAutotracking      |   true            |           
| screenViewAutotracking     |   true            |           

> | sessionContext      |         true      |        

Reviewer 3: Can we add a info block stating that we're internally ignoring all fields other than mentioned in the table.
I mean even if the user do something like this:
`val trackerConfiguration = TrackerConfiguration()
  .base64Encoding(false)`

We'll internally ignore `base64Encoding` value. 
Note: This kind of methods are kept for easy migration and they don't play a role in any way for tracking events. It doesn't matter user uses them or not there value will be ignored internally.

```
</span>
      </TabPanel>
      <TabPanel>
<span>

```swift
let trackerConfig = TrackerConfiguration()
  .base64Encoding(false)
  .logLevel(.debug)
  .deepLinkContext(true)
  .applicationContext(true)
  .platformContext(true)
  .geoLocationContext(true)
  .lifecycleAutotracking(true)
  .diagnosticAutotracking(true)
  .screenViewAutotracking(true)
  .screenContext(true)
  .applicationContext(true)
  .exceptionAutotracking(true)
  .installAutotracking(true)
  .userAnonymisation(false)
  .appId(APP_ID)
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>

Snowplow automatically captures and tracks the following data. Refer to the <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/#auto-tracked-events-and-entities">Auto-tracked events and entities</a> section for more information.

| Variable name | Default value    | 
| :---------- | :------------ | 
| logLevel     |        LogLevel.OFF       |           
| lifecycleAutotracking      |   true            |           
| screenViewAutotracking     |   true            |           
| sessionContext      |         true      |

<h3>SubjectConfiguration</h3>

> The <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/"><code class="inline-code">SubjectConfiguration</code></a> class can be used to capture basic information about the user and the app which is attached on all events as context entity.

Reviewer 2: ```suggestion
The <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/"><code class="inline-code">SubjectConfiguration</code></a> class can be used to capture basic information about the user which is attached on all events as context entity.
```

<div class="warningBlock">
Note that <code class="inline-code">userId</code> is a mandatory field. If not provided, the <code class="inline-code">identify</code> call is ignored.
</div>

<Tabs>
  <TabList>
    <Tab>Objective-C</Tab>
    <Tab>Swift</Tab>
  </TabList>
    <TabPanels>
      <TabPanel>
<span>

```objectivec
RSSubjectConfiguration * subjectConfig = [
  [RSSubjectConfiguration alloc] init
];
[subjectConfig userId: @ "user_id"];
[subjectConfig traits: @ {
  @ "key_1": @ "value_1", @ "key_2": @20, @ "key_3": @YES
}];
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```swift
let subjectConfig = SubjectConfiguration()
  .userId("user_id")
  .traits(["key_1": "value_1", "key_2": 20, "key_3": true])
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>   
      </TabPanel>
  </TabPanels>
</Tabs>

## Updating class names

The below table lists the corresponding class names in Snowplow and RudderStack which need to be updated based on your platform:

| Snowplow    |     |   RudderStack    |  |  |
| :------------- | :-------------- |:------------- | :-------------- |:-------------- |

> |           | **Java** | **Kotlin** | **Objective-C** | **Swift** |

Reviewer 3: We need to add one more class `LogLevel` in the table. Corresponding class name for all other column will also be same (i.e., LogLevel).

| `TrackerConfiguration` | `TrackerConfiguration` | `TrackerConfiguration` | `RSTrackerConfiguration` | `TrackerConfiguration` |
| `SessionConfiguration` | `SessionConfiguration` | `SessionConfiguration` | `RSSessionConfiguration` | `SessionConfiguration` |
| `SubjectConfiguration` | `SubjectConfiguration` | `SubjectConfiguration` | `RSSubjectConfiguration` | `SubjectConfiguration` |
| `Structured` | `Structured` | `Structured` | `RSStructured` | `Structured` |
| `ScreenView` | `ScreenView` | `ScreenView` | `RSScreenView` | `ScreenView` |
| `Background` | `Background` | `Background` | `RSBackground` | `Background` |
| `Foreground` | `Foreground` | `Foreground` | `RSForeground` | `Foreground` |
| `SelfDescribing` | `SelfDescribing` | `SelfDescribing` | `RSSelfDescribing` | `SelfDescribing` |
| `SelfDescribingJson` | `SelfDescribingJson` | `SelfDescribingJson` | `RSSelfDescribingJson` | `SelfDescribingJson` |
| `TimeMeasure` | `TimeMeasure` | `TimeMeasure` | `N/A` | `N/A` |
| `Snowplow` | `RSTracker` | `RSTracker` | `RSTracker` | `RSTracker` |

## Sending event data

> You can migrate your existing events from RudderStack to Snowplow as described below:

Reviewer 2: ```suggestion
You can migrate your existing events from Snowplow to RudderStack as described below:
```

### Identifying users

<Tabs>
  <TabList>
    <Tab>iOS/Android</Tab>
    <Tab>JavaScript</Tab>
  </TabList>
    <TabPanels>
      <TabPanel>
Snowplow's <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/"><code class="inline-code">SubjectConfiguration</code></a> class can be used to identify the users. Refer to the <Link to="/user-guides/migration-guides/snowplow-migration-guide/#setting-the-configuration-objects">Setting the configuration objects</Link> section for more information.
      </TabPanel>
      <TabPanel>

Snowplow supports identifying a user by passing the <code class="inline-code">setUserId</code> property. A Snowplow event including the <code class="inline-code">setUserId</code> property triggers the <Link to="/event-spec/standard-events/identify"><code class="inline-code">identify</code></Link> call in RudderStack JavaScript SDK.

<div class="infoBlock">

> RudderStack supports only the v2 Snowplow JavaScript tracker.

Reviewer 4: This `setUserId` method is only supported in the Snowplow v2 JS tracker.
The rest of the events are from the v3 JS tracker.
So, `setUserId` is the only method supported by Rudderstack from the v2 version of the Snowplow tracker.
@ishachopra0401 , please rephrase the sentence.

</div>

In the below sample call, <code class="inline-code">john.doe@email.com</code> is the <code class="inline-code">userId</code> and the <code class="inline-code">firstName</code>, <code class="inline-code">lastName</code>, and <code class="inline-code">city</code> are the user traits.
<span>

```javascript
rs.snowplowAdapter('setUserId', 'john.doe@email.com', {
  firstName: 'John',
  lastName: 'Doe',
  city: 'Barcelona',
});
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>

> ### Tracking users

Reviewer 2: ```suggestion
### Tracking events
```

> RudderStack SDKs support the following types of Snowplow events to track users:

Reviewer 2: ```suggestion
RudderStack SDKs support the following types of Snowplow events:
```

<Tabs>

>   <TabList>

Reviewer 2: Let's have two tabs. iOS/Android & JavaScript
Then show 4 tabs(Objective C, Swift, Java, Kotlin) under the code snippets of iOS/Android.

      revenue: 30,
      currency: 'USD',
      user_actual_id: 12345,
    },
  },
}); 
```
</span>
In the above example, the <code class="inline-code">trackSelfDescribingEvent</code> method tracks the <code class="inline-code">Order Completed</code> event along with other information like <code class="inline-code">revenue</code>, <code class="inline-code">currency</code>, and <code class="inline-code">user_actual_id</code>, etc.
<br/><br/>
A sample <code class="inline-code">trackStructEvent</code> call is shown below:
<span>

```javascript
rs.snowplowAdapter('trackStructEvent', {
  action: 'order completed',
  category: 'FCW',
  label: 'Sample label',
  property: 'Some property',
  value: 40.0,
});
```
</span>
In the above example, the <code class="inline-code">trackStructEvent</code> method tracks the <code class="inline-code">Order Completed</code> event along with other information like <code class="inline-code">label</code>, <code class="inline-code">property</code>, and <code class="inline-code">value</code>, etc.

<div class="infoBlock">
The <code class="inline-code">action</code> field is mandatory.
</div>
      </TabPanel>
    </TabPanels>
</Tabs>

> ### Tracking users' activities

Reviewer 2: Let's follow the same:
```
Let's have two tabs. iOS/Android & JavaScript
Then show 4 tabs(Objective C, Swift, Java, Kotlin) under the code snippets of iOS/Android.
```

<Tabs>
  <TabList>
    <Tab>Android</Tab>
    <Tab>iOS</Tab>
    <Tab>JavaScript</Tab>
  </TabList>
    <TabPanels>
      <TabPanel>
Snowplow's <a href="https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/mobile-trackers/tracking-events/#creating-a-structured-event"><code class="inline-code">ScreenView</code></a> class captures whenever a new screen is loaded. A Snowplow event including the <code class="inline-code">ScreenView</code> class triggers the <Link to="/event-spec/standard-events/screen/">screen</Link> call in the RudderStack Android SDK.
<br/><br/>

>  RudderStack maps the <code class="inline-code">previousName</code> to RudderStack event name and the rest of the properties, like <code class="inline-code">screenId</code>, <code class="inline-code">previousType</code>, <code class="inline-code">previousId</code>, etc. to the RudderStack properties.

Reviewer 2: ```suggestion
 RudderStack maps the <code class="inline-code">name</code> to RudderStack event name and the rest of the properties, like <code class="inline-code">screenId</code>, <code class="inline-code">previousType</code>, <code class="inline-code">previousId</code>, etc. to the RudderStack properties.
```
